src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -430,44 +462,59 @@ protected void InitializeReferences()
 
             _refsInitialized = true;
             _runnerref = new ExclusiveReference();
-            _replref = new WeakReference<RegexReplacement>(null);
+            ReplRef = new WeakReference<RegexReplacement>(null);
         }
 
         /// <summary>
-        /// Internal worker called by all the public APIs
+        /// Internal worker called by all the public APIs. Accepts both a Memory and Span. Usually one of these 
+        /// two is empty depending on the scenario. Unidirectional APIs that are unidirectional like IsMatch, 
+        /// Replace (with no MatchEvaluator) and Split should pass the input text as a Span (with an empty Memory).
+        /// Bidirectional APIs where the input needs to be passed around i.e. Match and Matches, should pass the input
+        /// text as a Memory (with an empty Span). Replace with an MatchEvaluator is special as it's unidirectional but
+        /// the evaluator can access the Match object which is the internal state and contains the input text.
         /// </summary>
-        /// <returns></returns>
-        internal Match Run(bool quick, int prevlen, string input, int beginning, int length, int startat)
+        internal Match Run(bool quick, int prevlen, in MemoryOrPinnedSpan<char> mem, ReadOnlySpan<char> span, int beginning, int length, int startat)
         {
+            // If a non empty Span is passed, use it and avoid Span creation from Memory.
+            ReadOnlySpan<char> input = !span.IsEmpty ? span : mem.Span;
+
             if (startat < 0 || startat > input.Length)
                 throw new ArgumentOutOfRangeException(nameof(startat), SR.BeginIndexNotNegative);
-
             if (length < 0 || length > input.Length)
                 throw new ArgumentOutOfRangeException(nameof(length), SR.LengthNotNegative);
 
-            // There may be a cached runner; grab ownership of it if we can.
-            RegexRunner runner = _runnerref.Get();
+            Match match;
 
-            // Create a RegexRunner instance if we need to
-            if (runner == null)
+            // Interpreted
+            if (factory == null)
             {
-                // Use the compiled RegexRunner factory if the code was compiled to MSIL
-                if (factory != null)
-                    runner = factory.CreateInstance();
-                else
-                    runner = new RegexInterpreter(_code, UseOptionInvariant() ? CultureInfo.InvariantCulture : CultureInfo.CurrentCulture);
-            }
+                RegexInterpreter interpretedRunner = _runnerref.Get() as RegexInterpreter ??
```

Comment:
What happens if _runnerref.Get() returns something non-null other than RegexInterpreter?  We don't end up releasing it.